### PR TITLE
DAOS-16365 client: intercept dlsym() and zeInit() to avoid nested call

### PR DESCRIPTION
**Pull request by wiliamhuang**
libfabric loads libze_loader.so which calls zeInit(). We observed deadlock due to nested calls when daos_init() is called inside zeInit(). We intercept dlsym() and zeInit() to avoid calling daos_init() inside zeInit(). dlsym(RTLD_NEXT, ) checks returning address to determine caller's module. To maintain expected behavior of dlsym(RTLD_NEXT, ) with our interception, new_dlsym() is implemented with assembly code to use jmp instruction instead of call. dlsym() has been moved from libdl.so to libc.so since version 2.34.

Features: pil4dfs

Required-githooks: true

### Before requesting gatekeeper:

* [ ] Two review approvals and any prior change requests have been resolved.
* [ ] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [ ] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [ ] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [ ] Any tests skipped by the ticket being addressed have been run and passed in the PR.

### Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate owners.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/client/dfuse/pil4dfs/int_dfs.c [hunk 1]
```diff
@@ -1020,6 +1026,138 @@ consume_low_fd(void)
...
+/* c code for other architecture. caller info could be wrong inside libc dlsym() when handle is set
+ * RTLD_NEXT. Assembly version implementation similar to above is needed to fix the issue by using
+ * jump instead of call instruction. 
+ */
+static void *
+new_dlsym_c(void *handle, const char *symbol)
+{
+	if (!d_hook_enabled)
+		goto org_dlsym;
+	printf("Inside my dlsym().\n");
```
Comment (wiliamhuang) [inline src/client/dfuse/pil4dfs/int_dfs.c:1144]: This line was for debugging. I will remove it later. This function is only compiled on non-x86_64 architecture.

File: src/client/dfuse/pil4dfs/hook.c [hunk 2]
```diff
@@ -89,10 +89,16 @@ static uint64_t lib_base_addr[MAX_NUM_LIB];
 /* List of names of loaded libraries */
 static char   **lib_name_list;
 
+/* we set libc_version as DEFAULT_LIBC_VER when we fail to determine it from libc.so */
+#define DEFAULT_LIBC_VER 2.28
```
Comment (mchaarawi) [inline src/client/dfuse/pil4dfs/hook.c:93]: this does not sound correct to me and it looks a bit hacky

Comment (wiliamhuang) [reply src/client/dfuse/pil4dfs/hook.c:93]: ok. I will check whether there exists an API to query libc version.

**Comment by wiliamhuang** [comment]
> im still not convinced this is something safe to do here.

We intercept many functions in our libraries. Do you have concern about intercepting dlsym()? If we encounter any issue, we could resolve it later. From my understanding, applications calling dlsym() should behave same as expected except dlsym(RTLD_NEXT, "zeInit") because we want to intercept zeInit() here. libfabric calls dlsym(libze_loader_module, "zeInit") to query the address of zeInit().

**Comment by wiliamhuang** [comment]
> Can you point me to anywhere in our client (pil4dfs or ioil) where we did something like this before?

All functions intercepted with register_a_hook() are based on trampoline method in libpil4dfs. 

> yes. What you are pushing here does not look safe to me and looks like you have added a bunch of hacks or workarounds in this patch to get around what libfabric is doing. we should not be doing that.

One hacky part is the assembly implementation of new_dlsym(). This is needed for correct behavior for dlsym(RTLD_NEXT, ). Other workarounds are for compatibilities for compilers/compiling flags we use.
dlsym() has been moved from libdl.so to libc.so since version 2.34. This also makes the interception of dlsym() a little more complicated.

> Sorry but this is not acceptable statement for a production code base.
ok. Then shall we run more tests to verify whether it is stable enough?

**Comment by mchaarawi** [comment]
> > Can you point me to anywhere in our client (pil4dfs or ioil) where we did something like this before?
> All functions intercepted with register_a_hook() are based on trampoline method in libpil4dfs.

which ones are we messing with the assembly code of the function?

> 
> > yes. What you are pushing here does not look safe to me and looks like you have added a bunch of hacks or workarounds in this patch to get around what libfabric is doing. we should not be doing that.
> 
> One hacky part is the assembly implementation of new_dlsym(). This is needed for correct behavior for dlsym(RTLD_NEXT, ). Other workarounds are for compatibilities for compilers/compiling flags we use. dlsym() has been moved from libdl.so to libc.so since version 2.34. This also makes the interception of dlsym() a little more complicated.

exactly.. these workarounds do not seem to be an acceptable thing to have in general.

> 
> > Sorry but this is not acceptable statement for a production code base.
>ok. Then shall we run more tests to verify whether it is stable enough?

my opinion is to leave this for now. the MPI_init() solution is working fine now?

File: src/client/dfuse/pil4dfs/int_dfs.c [hunk 3]
```diff
@@ -1020,6 +1026,138 @@ consume_low_fd(void)
...
+#if defined(__x86_64__)
+/* This is used to work around compiling warning and limitations of using asm function. */
+static void *
+query_new_dlsym_addr(void *addr)
+{
+	int i;
+
+	/* assume little endian */
+	for (i = 0; i < 64; i++) {
+		if (*((int *)(addr + i)) == 0x56579090) {
```
[the remark this answers is not in the document]
Comment (wiliamhuang) [reply src/client/dfuse/pil4dfs/int_dfs.c:1056]: You are right. Some comments could help here.

File: src/client/dfuse/pil4dfs/hook.c (hunk 2 above)
Comment (wiliamhuang) [reply src/client/dfuse/pil4dfs/hook.c:93]: Function gnu_get_libc_version() could be used to query glibc version.

**Comment by wiliamhuang** [comment]
All CI tests passed.

File: src/client/dfuse/pil4dfs/int_dfs.c [hunk 4]
```diff
@@ -1020,6 +1026,138 @@ consume_low_fd(void)
...
+{
+	return (void *)zeInit;
+}
+
+__attribute__((aligned(16))) static void
+new_dlsym_marker(void)
+{
+}
+
+/*__attribute__((naked,unused)) int new_dlsym(void *handle, const char *symbol) */
```
[the remark this answers is not in the document]
Comment (wiliamhuang) [reply src/client/dfuse/pil4dfs/int_dfs.c]: Fixed as you suggested. Thank you!

File: src/client/dfuse/pil4dfs/int_dfs.c (hunk 3 above)
[the remark this answers is not in the document]
Comment (wiliamhuang) [reply src/client/dfuse/pil4dfs/int_dfs.c:1111]: Revised as you suggested. Thank you very much!

File: src/client/dfuse/pil4dfs/int_dfs.c (hunk 1 above)
Comment (wiliamhuang) [reply src/client/dfuse/pil4dfs/int_dfs.c:1201]: Forgot to remove this line in recent update. It will be removed in next commit.

**Review by mchaarawi — COMMENTED** [review]
looks fine to me

File: src/client/dfuse/pil4dfs/hook.c [hunk 5]
```diff
@@ -290,19 +296,17 @@ determine_lib_path(void)
...
 	}
 
-	if (lib_ver_str[0]) {
-		/* with version in name */
-		rc = asprintf(&path_libpthread, "%s/libpthread-%s.so", lib_dir_str, lib_ver_str);
-	} else {
-		rc = asprintf(&path_libpthread, "%s/libpthread.so.0", lib_dir_str);
-	}
+	/* with version in name */
+	rc = asprintf(&path_libpthread, "%s/libpthread-%s.so", lib_dir_str, libc_version_str);
```
Comment (jolivier23) [inline src/client/dfuse/pil4dfs/hook.c:309]: reason to not use D_ASPRINTF ?

File: src/client/dfuse/pil4dfs/int_dfs.c [hunk 6]
```diff
@@ -1074,6 +1078,143 @@ PMPI_Init(int *argc, char ***argv)
...
+_Pragma("GCC optimize(\"-O0\")")
+static char str_zeinit[] = "zeInit";
+
+static int
+is_hook_enabled(void)
+{
+	return (d_hook_enabled ? (1) : (0));
+}
+
+/* This wrapper function is introduced to avoid compiling issue with Intel-C on Leap 15.5 */
```
Comment (jolivier23) [inline src/client/dfuse/pil4dfs/int_dfs.c:1135]: if this is a bug specific to Intel-C, should we limit the implementation to that compiler?

File: src/client/dfuse/pil4dfs/hook.c (hunk 5 above)
Comment (wiliamhuang) [reply src/client/dfuse/pil4dfs/hook.c:309]: You are right. I will replace it with D_ASPRINTF in next commit. Thank you!

File: src/client/dfuse/pil4dfs/int_dfs.c (hunk 6 above)
Comment (wiliamhuang) [reply src/client/dfuse/pil4dfs/int_dfs.c:1135]: We only see issues in Intel C in CI. We do not know whether other compilers we did not test have same issues. The workaround here might provide better compatibility.

**Review by knard-intel — APPROVED** [review]

**Review by mchaarawi — APPROVED** [review]

**Pull request merged**